File: apps/api/app/services/n8n_node_registry.py
"""n8n node registry — metadata, aliases, and official-type detection."""

from __future__ import annotations

import re
from typing import Any

N8N_OFFICIAL_PREFIXES: tuple[str, ...] = (
    "n8n-nodes-base.",
    "@n8n/n8n-nodes-langchain.",
    "n8n-nodes-langchain.",
)

# Deprecated n8n node types → current equivalents.
DEPRECATED_ALIASES: dict[str, str] = {
    "n8n-nodes-base.function": "n8n-nodes-base.code",
    "n8n-nodes-base.functionItem": "n8n-nodes-base.code",
    "n8n-nodes-base.venafi": "n8n-nodes-base.venafiTlsProtectCloud",
    "n8n-nodes-base.venafiTrigger": "n8n-nodes-base.venafiTlsProtectCloudTrigger",
}

# Curated metadata for common node types (extensible without touching core import logic).
KNOWN_NODES: dict[str, dict[str, str]] = {
    "n8n-nodes-base.manualTrigger": {"display_name": "Manual Trigger", "category": "Trigger"},
    "n8n-nodes-base.start": {"display_name": "Start", "category": "Trigger"},
    "n8n-nodes-base.webhook": {"display_name": "Webhook", "category": "Trigger"},
    "n8n-nodes-base.cron": {"display_name": "Cron", "category": "Trigger"},
    "n8n-nodes-base.schedule": {"display_name": "Schedule", "category": "Trigger"},
    "n8n-nodes-base.formTrigger": {"display_name": "Form Trigger", "category": "Trigger"},
    "n8n-nodes-base.emailReadImap": {"display_name": "Email Trigger (IMAP)", "category": "Trigger"},
    "n8n-nodes-base.if": {"display_name": "IF", "category": "Logic"},
    "n8n-nodes-base.switch": {"display_name": "Switch", "category": "Logic"},
    "n8n-nodes-base.merge": {"display_name": "Merge", "category": "Logic"},
    "n8n-nodes-base.splitInBatches": {"display_name": "Split In Batches", "category": "Logic"},
    "n8n-nodes-base.wait": {"display_name": "Wait", "category": "Logic"},
    "n8n-nodes-base.noOp": {"display_name": "No Operation", "category": "Core"},
    "n8n-nodes-base.set": {"display_name": "Set", "category": "Core"},
    "n8n-nodes-base.code": {"display_name": "Code", "category": "Core"},
    "n8n-nodes-base.httpRequest": {"display_name": "HTTP Request", "category": "HTTP"},
    "n8n-nodes-base.respondToWebhook": {"display_name": "Respond to Webhook", "category": "HTTP"},
    "n8n-nodes-base.aggregate": {"display_name": "Aggregate", "category": "Core"},
    "n8n-nodes-base.gmail": {"display_name": "Gmail", "category": "Communication"},
    "n8n-nodes-base.googleSheets": {"display_name": "Google Sheets", "category": "Google Workspace"},
    "n8n-nodes-base.googleDrive": {"display_name": "Google Drive", "category": "Google Workspace"},
    "n8n-nodes-base.googleDocs": {"display_name": "Google Docs", "category": "Google Workspace"},
    "n8n-nodes-base.googleCalendar": {"display_name": "Google Calendar", "category": "Google Workspace"},
    "n8n-nodes-base.slack": {"display_name": "Slack", "category": "Communication"},
    "n8n-nodes-base.discord": {"display_name": "Discord", "category": "Communication"},
    "n8n-nodes-base.telegram": {"display_name": "Telegram", "category": "Communication"},
    "n8n-nodes-base.notion": {"display_name": "Notion", "category": "Productivity"},
    "n8n-nodes-base.github": {"display_name": "GitHub", "category": "Development"},
    "n8n-nodes-base.postgres": {"display_name": "Postgres", "category": "Database"},
    "n8n-nodes-base.mysql": {"display_name": "MySQL", "category": "Database"},
    "n8n-nodes-base.mongodb": {"display_name": "MongoDB", "category": "Database"},
    "n8n-nodes-base.redis": {"display_name": "Redis", "category": "Database"},
    "n8n-nodes-base.openAi": {"display_name": "OpenAI", "category": "AI"},
    "n8n-nodes-base.anthropic": {"display_name": "Anthropic", "category": "AI"},
    "n8n-nodes-base.workflow": {"display_name": "Execute Workflow", "category": "Flow"},
    "n8n-nodes-base.executeWorkflow": {"display_name": "Execute Workflow", "category": "Flow"},
    "n8n-nodes-base.awsS3": {"display_name": "AWS S3", "category": "AWS"},
    "n8n-nodes-base.awsLambda": {"display_name": "AWS Lambda", "category": "AWS"},
    "n8n-nodes-base.microsoftOutlook": {"display_name": "Microsoft Outlook", "category": "Microsoft"},
    "n8n-nodes-base.microsoftTeams": {"display_name": "Microsoft Teams", "category": "Microsoft"},
    "@n8n/n8n-nodes-langchain.agent": {"display_name": "AI Agent", "category": "AI / LangChain"},
    "@n8n/n8n-nodes-langchain.lmChatOpenAi": {"display_name": "OpenAI Chat Model", "category": "AI / LangChain"},
    "@n8n/n8n-nodes-langchain.lmChatAnthropic": {"display_name": "Anthropic Chat Model", "category": "AI / LangChain"},
    "@n8n/n8n-nodes-langchain.lmChatGoogleGemini": {"display_name": "Google Gemini Chat Model", "category": "AI / LangChain"},
    "@n8n/n8n-nodes-langchain.toolWorkflow": {"display_name": "Workflow Tool", "category": "AI / LangChain"},
    "@n8n/n8n-nodes-langchain.manualTrigger": {"display_name": "Manual Trigger", "category": "Trigger"},
}

# Keys preserved on every imported n8n node for lossless round-trip.
N8N_NODE_PRESERVE_KEYS: tuple[str, ...] = (
    "id",
    "name",
    "type",
    "typeVersion",
    "parameters",
    "position",
    "credentials",
    "disabled",
    "notes",
    "webhookId",
    "continueOnFail",
    "retryOnFail",
    "maxTries",
    "waitBetweenTries",
    "alwaysOutputData",
    "executeOnce",
    "onError",
    "color",
    "pinData",
)

N8N_WORKFLOW_PRESERVE_KEYS: tuple[str, ...] = (
    "settings",
    "pinnedData",
    "meta",
    "staticData",
    "tags",
    "versionId",
    "hash",
)


def normalize_n8n_type(n8n_type: str) -> str:
    return DEPRECATED_ALIASES.get(n8n_type, n8n_type)


def is_official_n8n_type(n8n_type: str) -> bool:
    normalized = normalize_n8n_type(n8n_type)
    return any(normalized.startswith(prefix) for prefix in N8N_OFFICIAL_PREFIXES)


def _short_type(n8n_type: str) -> str:
    for prefix in N8N_OFFICIAL_PREFIXES:
        if n8n_type.startswith(prefix):
            return n8n_type[len(prefix) :]
    return n8n_type


def _camel_to_title(value: str) -> str:
    spaced = re.sub(r"([a-z0-9])([A-Z])", r"\1 \2", value)
    spaced = re.sub(r"([A-Z]+)([A-Z][a-z])", r"\1 \2", spaced)
    return spaced.replace("_", " ").replace("-", " ").strip().title()


def _infer_category(n8n_type: str, short: str) -> str:
    lower = n8n_type.lower()
    short_lower = short.lower()
    if short_lower.endswith("trigger") or "trigger" in short_lower:
        return "Trigger"
    if "langchain" in lower or short_lower.endswith("tool"):
        return "AI / LangChain"
    if short_lower.startswith("google") or "googledocs" in short_lower or "googledrive" in short_lower:
        return "Google Workspace"
    if short_lower.startswith("microsoft") or short_lower in {"outlook", "teams", "onedrive"}:
        return "Microsoft"
    if short_lower.startswith("aws"):
        return "AWS"
    if short_lower in {"slack", "discord", "telegram", "twilio", "whatsapp", "gmail", "sendgrid"}:
        return "Communication"
    if any(k in short_lower for k in ("postgres", "mysql", "mongo", "redis", "sqlite", "supabase", "airtable")):
        return "Database"
    if any(k in short_lower for k in ("openai", "anthropic", "gemini", "huggingface", "cohere", "mistral")):
        return "AI"
    if short_lower in {"if", "switch", "merge", "splitinbatches", "wait", "filter", "compare"}:
        return "Logic"
    if short_lower in {"httprequest", "respondtowebhook", "webhook"}:
        return "HTTP"
    if short_lower in {"notion", "airtable", "clickup", "jira", "linear", "asana", "trello"}:
        return "Productivity"
    if short_lower in {"github", "gitlab", "bitbucket", "jira"}:
        return "Development"
    if short_lower in {"workflow", "executeworkflow", "executeworkflowtrigger"}:
        return "Flow"
    if is_official_n8n_type(n8n_type):
        return "Core"
    return "Community"


def get_node_metadata(n8n_type: str) -> dict[str, Any]:
    normalized = normalize_n8n_type(n8n_type)
    known = KNOWN_NODES.get(normalized) or KNOWN_NODES.get(n8n_type)
    short = _short_type(normalized)
    display_name = known["display_name"] if known else _camel_to_title(short)
    category = known["category"] if known else _infer_category(normalized, short)
    return {
        "n8n_type": n8n_type,
        "normalized_type": normalized,
        "display_name": display_name,
        "category": category,
        "official": is_official_n8n_type(n8n_type),
        "short_type": short,
    }


def build_node_snapshot(n8n_node: dict[str, Any]) -> dict[str, Any]:
    """Extract a lossless n8n node snapshot from raw workflow JSON."""
    snapshot: dict[str, Any] = {}
    for key in N8N_NODE_PRESERVE_KEYS:
        if key in n8n_node:
            snapshot[key] = n8n_node[key]
    # Preserve any extra keys n8n may add in future versions.
    for key, value in n8n_node.items():
        if key not in snapshot:
            snapshot[key] = value
    if "parameters" not in snapshot:
        snapshot["parameters"] = {}
    return snapshot


def extract_workflow_extras(n8n_data: dict[str, Any]) -> dict[str, Any]:
    extras: dict[str, Any] = {}
    for key in N8N_WORKFLOW_PRESERVE_KEYS:
        if key in n8n_data:
            extras[key] = n8n_data[key]
    return extras

File: apps/web/src/lib/n8n-node-registry.ts
/**
 * n8n node registry — display metadata and official-type detection.
 * Mirrors apps/api/app/services/n8n_node_registry.py for client-side rendering.
 */

export const N8N_OFFICIAL_PREFIXES = [
  "n8n-nodes-base.",
  "@n8n/n8n-nodes-langchain.",
  "n8n-nodes-langchain.",
] as const;

export const DEPRECATED_ALIASES: Record<string, string> = {
  "n8n-nodes-base.function": "n8n-nodes-base.code",
  "n8n-nodes-base.functionItem": "n8n-nodes-base.code",
};

export interface N8nNodeMetadata {
  n8nType: string;
  normalizedType: string;
  displayName: string;
  category: string;
  official: boolean;
  shortType: string;
}

const KNOWN_NODES: Record<string, { displayName: string; category: string }> = {
  "n8n-nodes-base.manualTrigger": { displayName: "Manual Trigger", category: "Trigger" },
  "n8n-nodes-base.webhook": { displayName: "Webhook", category: "Trigger" },
  "n8n-nodes-base.if": { displayName: "IF", category: "Logic" },
  "n8n-nodes-base.set": { displayName: "Set", category: "Core" },
  "n8n-nodes-base.aggregate": { displayName: "Aggregate", category: "Core" },
  "n8n-nodes-base.respondToWebhook": { displayName: "Respond to Webhook", category: "HTTP" },
  "n8n-nodes-base.googleSheets": { displayName: "Google Sheets", category: "Google Workspace" },
  "n8n-nodes-base.googleSheetsTool": { displayName: "Google Sheets Tool", category: "AI / LangChain" },
  "n8n-nodes-base.googleDocsTool": { displayName: "Google Docs Tool", category: "AI / LangChain" },
  "n8n-nodes-base.googleDriveTool": { displayName: "Google Drive Tool", category: "AI / LangChain" },
  "n8n-nodes-base.slackTool": { displayName: "Slack Tool", category: "AI / LangChain" },
  "n8n-nodes-base.telegramTool": { displayName: "Telegram Tool", category: "AI / LangChain" },
  "n8n-nodes-base.slack": { displayName: "Slack", category: "Communication" },
  "n8n-nodes-base.workflow": { displayName: "Execute Workflow", category: "Flow" },
  "n8n-nodes-base.openAi": { displayName: "OpenAI", category: "AI" },
  "@n8n/n8n-nodes-langchain.agent": { displayName: "AI Agent", category: "AI / LangChain" },
};

export function normalizeN8nType(n8nType: string): string {
  return DEPRECATED_ALIASES[n8nType] ?? n8nType;
}

export function isOfficialN8nType(n8nType: string): boolean {
  const normalized = normalizeN8nType(n8nType);
  return N8N_OFFICIAL_PREFIXES.some((prefix) => normalized.startsWith(prefix));
}

function shortType(n8nType: string): string {
  for (const prefix of N8N_OFFICIAL_PREFIXES) {
    if (n8nType.startsWith(prefix)) return n8nType.slice(prefix.length);
  }
  return n8nType;
}

function camelToTitle(value: string): string {
  return value
    .replace(/([a-z0-9])([A-Z])/g, "$1 $2")
    .replace(/([A-Z]+)([A-Z][a-z])/g, "$1 $2")
    .replace(/[_-]/g, " ")
    .trim()
    .replace(/\b\w/g, (c) => c.toUpperCase());
}

function inferCategory(n8nType: string, short: string): string {
  const lower = n8nType.toLowerCase();
  const s = short.toLowerCase();
  if (s.endsWith("trigger") || s.includes("trigger")) return "Trigger";
  if (lower.includes("langchain") || s.endsWith("tool")) return "AI / LangChain";
  if (s.startsWith("google")) return "Google Workspace";
  if (s.startsWith("microsoft") || s === "outlook" || s === "teams") return "Microsoft";
  if (s.startsWith("aws")) return "AWS";
  if (["slack", "discord", "telegram", "gmail", "twilio"].includes(s)) return "Communication";
  if (["postgres", "mysql", "mongodb", "redis", "sqlite"].some((k) => s.includes(k))) return "Database";
  if (["openai", "anthropic", "gemini", "huggingface"].some((k) => s.includes(k))) return "AI";
  if (["if", "switch", "merge", "wait", "filter"].includes(s)) return "Logic";
  if (["httprequest", "respondtowebhook"].includes(s)) return "HTTP";
  if (isOfficialN8nType(n8nType)) return "Core";
  return "Community";
}

export function getN8nNodeMetadata(n8nType: string): N8nNodeMetadata {
  const normalized = normalizeN8nType(n8nType);
  const known = KNOWN_NODES[normalized] ?? KNOWN_NODES[n8nType];
  const short = shortType(normalized);
  return {
    n8nType,
    normalizedType: normalized,
    displayName: known?.displayName ?? camelToTitle(short),
    category: known?.category ?? inferCategory(normalized, short),
    official: isOfficialN8nType(n8nType),
    shortType: short,
  };
}

/** Node types with native AgentFlow editor + local execution. */
export const NATIVE_MAPPED_TYPES = new Set([
  "n8n-nodes-base.manualTrigger",
  "n8n-nodes-base.start",
  "n8n-nodes-base.webhook",
  "n8n-nodes-base.if",
  "n8n-nodes-base.gmail",
  "n8n-nodes-base.googleSheets",
  "@n8n/n8n-nodes-langchain.agent",
  "n8n-nodes-base.openAi",
  "@n8n/n8n-nodes-langchain.lmChatOpenAi",
  "n8n-nodes-base.httpRequest",
  "n8n-nodes-base.noOp",
]);

export function isN8nCompatibleNode(n8nType: string): boolean {
  const normalized = normalizeN8nType(n8nType);
  return NATIVE_MAPPED_TYPES.has(normalized) || isOfficialN8nType(normalized);
}

export function getCategoryColor(category: string): string {
  const map: Record<string, string> = {
    Trigger: "bg-emerald-500/10 text-emerald-700 dark:text-emerald-400",
    Logic: "bg-amber-500/10 text-amber-700 dark:text-amber-400",
    Core: "bg-sky-500/10 text-sky-700 dark:text-sky-400",
    HTTP: "bg-blue-500/10 text-blue-700 dark:text-blue-400",
    "AI / LangChain": "bg-violet-500/10 text-violet-700 dark:text-violet-400",
    AI: "bg-violet-500/10 text-violet-700 dark:text-violet-400",
    "Google Workspace": "bg-green-500/10 text-green-700 dark:text-green-400",
    Communication: "bg-pink-500/10 text-pink-700 dark:text-pink-400",
    Database: "bg-cyan-500/10 text-cyan-700 dark:text-cyan-400",
    AWS: "bg-orange-500/10 text-orange-700 dark:text-orange-400",
    Microsoft: "bg-indigo-500/10 text-indigo-700 dark:text-indigo-400",
    Flow: "bg-teal-500/10 text-teal-700 dark:text-teal-400",
    Community: "bg-zinc-500/10 text-zinc-700 dark:text-zinc-400",
  };
  return map[category] ?? map.Core;
}
